### PR TITLE
Add callback failure warning email

### PR DESCRIPTION
**Pull request by whabanks**
# Summary | Résumé
This PR adds the groundwork for sending a warning email to service owners when their callback service is not working correctly. 

The actual email sending based on threshold is not yet implemented. It is non-trivial to ensure that we don't spam users with emails every time a retried callback reaches the max retries of 5. For the time being we will simply monitor how frequently this occurs while the remaining work is completed.

## Related Issues | Cartes liées
- https://github.com/cds-snc/notification-terraform/pull/1352
- https://github.com/cds-snc/notification-admin/pull/1855

# Test instructions | Instructions pour tester la modification

_TODO: Fill in test instructions for the reviewer._

# Release Instructions | Instructions pour le déploiement

None.

# Reviewer checklist | Liste de vérification du réviseur

- [ ] This PR does not break existing functionality.
- [ ] This PR does not violate GCNotify's privacy policies.
- [ ] This PR does not raise new security concerns. Refer to our GC Notify Risk Register document on our Google drive.
- [ ] This PR does not significantly alter performance.
- [ ] Additional required documentation resulting of these changes is covered (such as the README, setup instructions, a related ADR or the technical documentation).

> ⚠ If boxes cannot be checked off before merging the PR, they should be moved to the "Release Instructions" section with appropriate steps required to verify before release. For example, changes to celery code may require tests on staging to verify that performance has not been affected.

### CODE REVIEW

File: app/celery/service_callback_tasks.py [hunk 1]
```diff
@@ -86,5 +86,5 @@ def _send_data_to_service_callback_api(self, data, service_callback_url, token,
                 self.retry(queue=QueueNames.CALLBACKS_RETRY)
             except self.MaxRetriesExceededError:
                 current_app.logger.warning(
-                    "Retry: {function_name} has retried the max num of times for callback url {service_callback_url} and notification_id: {notification_id}"
+                    f"Retry: {function_name} has retried the max num of times for callback url {service_callback_url} and notification_id: {notification_id}"
```
Comment (whabanks) [inline app/celery/service_callback_tasks.py:89]: Turns out we have not been properly capturing callback failure logs for a while now.

File: app/config.py [hunk 2]
```diff
@@ -299,47 +299,48 @@ class Config(object):
...
-    EMAIL_2FA_TEMPLATE_ID = "299726d2-dba6-42b8-8209-30e1d66ea164"
-    EMAIL_MAGIC_LINK_TEMPLATE_ID = "6e97fd09-6da0-4cc8-829d-33cf5b818103"
-    NEW_USER_EMAIL_VERIFICATION_TEMPLATE_ID = "ece42649-22a8-4d06-b87f-d52d5d3f0a27"
-    PASSWORD_RESET_TEMPLATE_ID = "474e9242-823b-4f99-813d-ed392e7f1201"
-    FORCED_PASSWORD_RESET_TEMPLATE_ID = "e9a65a6b-497b-42f2-8f43-1736e43e13b3"
+    ACCOUNT_CHANGE_TEMPLATE_ID = "5b39e16a-9ff8-487c-9bfb-9e06bdb70f36"
     ALREADY_REGISTERED_EMAIL_TEMPLATE_ID = "0880fbb1-a0c6-46f0-9a8e-36c986381ceb"
+    APIKEY_REVOKE_TEMPLATE_ID = "a0a4e7b8-8a6a-4eaa-9f4e-9c3a5b2dbcf3"
+    BRANDING_REQUEST_TEMPLATE_ID = "7d423d9e-e94e-4118-879d-d52f383206ae"
+    CALLBACK_FAILURE_TEMPLATE_ID = "d8d580f4-86b3-4ba4-9d7c-263a630af354"
```
Comment (whabanks) [inline app/config.py:306]: Amongst the rearranging, this env var is the one I added.

File: migrations/versions/0453_add_callback_failure_email.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            "",
+            "The GC Notify team",
+            "[[/en]]",
+            "",
+            "---",
+            "",
+            "[[fr]]",
+            "Bonjour ((name)),",
+            "",
+            "Les rappels pour « ((service_name)) » ne fonctionnent pas. Cela pourrait signifier que :" "",
```
[the remark this answers is not in the document]
Comment (whabanks) [reply migrations/versions/0453_add_callback_failure_email.py:56]: Fixed.

File: app/celery/service_callback_tasks.py [hunk 4]
```diff
@@ -80,11 +84,73 @@ def _send_data_to_service_callback_api(self, data, service_callback_url, token,
...
+        #         send_email_callback_failure_email(current_app.service)
+
         # Retry if the response status code is server-side or 429 (too many requests).
         if not isinstance(e, HTTPError) or e.response.status_code >= 500 or e.response.status_code == 429:
             try:
                 self.retry(queue=QueueNames.CALLBACKS_RETRY)
             except self.MaxRetriesExceededError:
                 current_app.logger.warning(
-                    "Retry: {function_name} has retried the max num of times for callback url {service_callback_url} and notification_id: {notification_id}"
+                    f"Retry: {function_name} has retried the max num of times for callback url {service_callback_url} and notification_id: {notification_id} for service: {current_app.service.id}"
```
Comment (andrewleith) [inline app/celery/service_callback_tasks.py:102]: For the limits feature, we use an expiring redis key to ensure only one email is sent within a particular time period. For example, if they reach the threshold, we send the email then set this key in redis to true. Then if they hit the threshold again within the time period the redis key will still be true and we will check it before sending another email.

File: app/config.py (hunk 2 above)
Comment (andrewleith) [reply app/config.py:306]: Easier to read, thank you 🔤

**Review by andrewleith — COMMENTED** [review]
Looks good.  If our threshold is simply whenever `MaxRetriesExceededError` is thrown, I think we can use a redis key to ensure we only send one email per day. See comment below.

**Comment by jzbahrai** [comment]
We are not doing this,this is already in the PRD GC notify service

**Pull request closed without merging**